### ISSUE TITLE
Stack overflow in `Collect::needs_trace` for enum with recursion in first variant

### ISSUE THREAD

**Issue by lgoeldner** [kyren/gc-arena#120]
```rust
use std::marker::PhantomData;

use gc_arena::{rootless_arena, Collect, Gc};

#[derive(Collect)]
#[collect(no_drop)]
enum A<'gc> {
    V(Vec<A<'gc>>),
    Bar(PhantomData<&'gc ()>),
}

fn main() {
    rootless_arena(|mc| drop(Gc::new(mc, A::Bar(PhantomData))));
}
```

This causes a stack overflow for me. It seems to have something to do with the `impl Collect` for Collections, that seems to trigger an infinite loop. It works with every Collection that has indirection and the `Collect::needs_trace` is `T::needs_trace()`.

**Comment by kyren** [comment]
Yep, this definitely sounds like a bug in the released version of `gc-arena`. This can be worked around by manually implementing `Collect` for your type.

In the master branch right now, the bug still *sort of* exists but it is much less harmful because it fails at compile time, since `Collect::NEEDS_TRACE` is now an associated constant (it fails for me with something like `cycle detected when elaborating drops for ...`).

In the master branch this can be worked around for types that actually contain `Gc` pointers by changing the ordering of enum fields such that `NEEDS_TRACE` is set to `true || <some other expression>`, short-cutting the failure. If, like in your example, the type doesn't actually contain any `Gc` pointers, then this will never work, but in this case you can get around the problem entirely with `#[collect(require_static)]`.

What we could do to cover every case without requiring the user to re-order fields or variants is allow you to set `#[collect(always_needs_trace)]` which could replace the generated expression for `NEEDS_TRACE` with just `true`, which is always safe to do (but may be less efficient). This is still confusing and not entirely ideal, but I can't think of a better solution right now.
